File: common/src/main/java/net/shiroha233/roadweaver/features/path/pathlogic/core/Road.java
package net.shiroha233.roadweaver.features.path.pathlogic.core;

import net.minecraft.core.BlockPos;
import net.minecraft.resources.ResourceLocation;
import net.minecraft.server.level.ServerLevel;
import net.minecraft.util.RandomSource;
import net.minecraft.world.level.block.state.BlockState;
import net.shiroha233.roadweaver.config.ConfigService;
import net.shiroha233.roadweaver.config.PresetService;
import net.shiroha233.roadweaver.config.sub.RoadGenerationConfig;
import net.shiroha233.roadweaver.core.model.RoadData;
import net.shiroha233.roadweaver.core.model.RoadSegmentPlacement;
import net.shiroha233.roadweaver.core.model.RoadSpan;
import net.shiroha233.roadweaver.core.model.SpanType;
import net.shiroha233.roadweaver.core.model.StructureConnection;
import net.shiroha233.roadweaver.features.path.config.PathFeatureConfig;
import net.shiroha233.roadweaver.features.path.pathlogic.pathfinding.RoadPathCalculator;
import net.shiroha233.roadweaver.features.path.pathlogic.pathfinding.RoadPathCalculator.PathCalculationResult;
import net.shiroha233.roadweaver.pathfinding.PathSpanExtractor;
import net.shiroha233.roadweaver.pathfinding.cache.TerrainSamplingCache;
import net.shiroha233.roadweaver.pathfinding.terrain.PathTerrainField;
import net.shiroha233.roadweaver.pathfinding.terrain.region.CoarsePathCache;
import net.shiroha233.roadweaver.pathfinding.terrain.region.CoarseTerrainRegion;
import net.shiroha233.roadweaver.pathfinding.terrain.region.CoarseTerrainRegionRegistry;
import net.shiroha233.roadweaver.persistence.sharded.RoadShardStorage;
import net.shiroha233.roadweaver.planning.PlanningUtils;
import net.shiroha233.roadweaver.structures.precompute.RoadsideStructurePrecomputer;
import net.shiroha233.roadweaver.structures.precompute.RoadsideVillagePrecomputer;

import java.util.*;

/**
 * 道路生成器
 */
public final class Road {
    private final ServerLevel level;
    private final StructureConnection connection;
    private final PathFeatureConfig featureConfig;
    private final RoadGenerationConfig genConfig;

    public Road(ServerLevel level, StructureConnection connection, 
                PathFeatureConfig featureConfig, RoadGenerationConfig genConfig) {
        this.level = level;
        this.connection = connection;
        this.featureConfig = featureConfig;
        this.genConfig = genConfig;
    }
    
    @Deprecated
    public Road(ServerLevel level, StructureConnection connection, PathFeatureConfig config) {
        this(level, connection, config, RoadGenerationConfig.from(ConfigService.get()));
    }

    public RoadData generateRoad(int maxSteps) {
        RandomSource random = RandomSource.create();
        int width = genConfig.effectiveRoadWidth(getRandomWidth(random, featureConfig));
        boolean allowA = genConfig.allowArtificial();
        boolean allowN = genConfig.allowNatural();
        if (!allowA && !allowN) return null;
        
        int type = allowA && allowN ? (random.nextBoolean() ? 0 : 1) : (allowA ? 0 : 1);
        List<BlockState> materials;
        List<BlockState> slabMaterials;
        PresetService.RoadType presetType = (type == 0) ? PresetService.RoadType.ARTIFICIAL : PresetService.RoadType.NATURAL;
        if (presetType == PresetService.RoadType.ARTIFICIAL) {
            PresetService.PresetDef preset = PresetService.choosePreset(random, presetType);
            materials = PresetService.toBlockStatesFromIds(preset.materials());
            slabMaterials = PresetService.toBlockStatesFromIds(preset.slabMaterials());
        } else {
            materials = List.of();
            slabMaterials = List.of();
        }

        BlockPos rawStart = connection.from();
        BlockPos rawEnd = connection.to();
        
        TerrainSamplingCache cache = new TerrainSamplingCache();
        PathTerrainField terrain = null;
        try {
            // 优先使用规划阶段已计算好的粗路径（粗采样数据已在规划阶段释放）
            List<BlockPos> coarsePath = CoarsePathCache.get(level, connection);

            PathCalculationResult pathResult;
            if (coarsePath != null) {
                // 粗路径已就绪，直接进入精采样
                pathResult = RoadPathCalculator.calculateFromCoarsePath(
                        coarsePath, rawStart, rawEnd, width, level, maxSteps, cache, genConfig);
            } else {
                // 无缓存粗路径（如旧版兼容），退回完整两阶段寻路
                CoarseTerrainRegion coarseRegion = CoarseTerrainRegionRegistry.acquire(level, connection);
                pathResult = RoadPathCalculator.calculateAStarRoadPathDetailed(
                        rawStart, rawEnd, width, level, maxSteps, cache, genConfig, coarseRegion);
                if (coarseRegion != null) coarseRegion.dispose();
                CoarseTerrainRegionRegistry.release(level, connection);
            }

            List<RoadSegmentPlacement> rawSegments = pathResult.segments();
            if (rawSegments == null || rawSegments.size() < 5) return null;

            List<RoadSegmentPlacement> segments = StructureRoadOffsetService.trimPathNearStructure(
                    level, rawSegments, rawStart, rawEnd);
            if (segments == null || segments.size() < 5) return null;

            terrain = pathResult.terrain();
            List<RoadSpan> spans = PathSpanExtractor.extractSpans(
                    segments, level, cache, terrain, genConfig.bridgeMinWaterDepth());
            List<Integer> targetY = computeTargetY(level, segments, spans, cache, genConfig);
            long ownerA2d = PlanningUtils.pos2dKey(rawStart);
            long ownerB2d = PlanningUtils.pos2dKey(rawEnd);

            RoadData rd = new RoadData(width, type, materials, slabMaterials, segments, spans, targetY, ownerA2d, ownerB2d);
            RoadShardStorage.addRoad(level, rd);

            RoadsideStructurePrecomputer.precomputeStructures(level, segments, spans, width, cache, random, targetY);
            RoadsideVillagePrecomputer.precomputeVillages(level, segments, spans, width, cache, terrain, random, targetY);
            return rd;
        } finally {
            if (terrain != null) terrain.dispose();
            CoarsePathCache.remove(level, connection);
            cache.clear();
        }
    }

    private static int getRandomWidth(RandomSource rnd, PathFeatureConfig cfg) {
        return 0;
    }
    
    private static List<Integer> computeTargetY(ServerLevel level, List<RoadSegmentPlacement> segments, 
                                                   List<RoadSpan> spans, TerrainSamplingCache cache,
                                                   RoadGenerationConfig cfg) {
        int n = segments.size();
        List<BlockPos> centers = new ArrayList<>(n);
        for (RoadSegmentPlacement s : segments) centers.add(s.middlePos());

        boolean[] isBridge = new boolean[n];
        if (spans != null && !spans.isEmpty()) {
            Map<Long, Integer> indexMap = new HashMap<>();
            for (int i = 0; i < centers.size(); i++) indexMap.put(centers.get(i).asLong(), i);
            for (RoadSpan sp : spans) {
                if (sp.type() != SpanType.BRIDGE) continue;
                Integer si = indexMap.get(sp.start().asLong());
                Integer ei = indexMap.get(sp.end().asLong());
                if (si == null || ei == null) continue;
                int a = Math.max(0, Math.min(si, ei));
                int b = Math.min(n - 1, Math.max(si, ei));
                for (int k = a; k <= b; k++) isBridge[k] = true;
            }
        }

        int avg = Math.max(0, cfg.averagingRadius());
        int[] base = new int[n];
        for (int i = 0; i < n; i++) {
            int sum = 0, cnt = 0;
            int lo = Math.max(0, i - avg);
            int hi = Math.min(n - 1, i + avg);
            for (int j = lo; j <= hi; j++) {
                BlockPos sp = centers.get(j);
                int yTop = sp.getY();
                sum += yTop; cnt++;
            }
            base[i] = cnt > 0 ? (int) Math.round(sum / (double) cnt) : centers.get(i).getY();
        }

        if (!cfg.slopeLimitEnabled()) {
            List<Integer> out = new ArrayList<>(n);
            for (int v : base) out.add(v);
            return out;
        }

        int[] smoothed = base.clone();
        int i = 0;
        while (i < n) {
            while (i < n && isBridge[i]) i++;
            int s = i;
            while (i < n && !isBridge[i]) i++;
            int e = i - 1;
            if (s <= e) {
                int step2 = Math.max(0, Math.min(8, cfg.maxSlopeStepPerTwoSegments()));
                int halfLow = Math.max(0, step2 / 2);
                int halfHigh = Math.max(0, (step2 + 1) / 2);
                for (int ii = s + 1; ii <= e; ii++) {
                    int y = smoothed[ii];
                    if (ii == s + 1) {
                        int py = smoothed[ii - 1];
                        if (y > py + halfLow) y = py + halfLow;
                        if (y < py - halfLow) y = py - halfLow;
                    } else {
                        int py = smoothed[ii - 1];
                        if (y > py + halfHigh) y = py + halfHigh;
                        if (y < py - halfHigh) y = py - halfHigh;
                        int p2 = smoothed[ii - 2];
                        int hi = p2 + step2;
                        int lo = p2 - step2;
                        if (y > hi) y = hi;
                        if (y < lo) y = lo;
                    }
                    smoothed[ii] = y;
                }
                for (int ii = e - 1; ii >= s; ii--) {
                    int y = smoothed[ii];
                    if (ii == e - 1) {
                        int ny = smoothed[ii + 1];
                        if (y > ny + halfLow) y = ny + halfLow;
                        if (y < ny - halfLow) y = ny - halfLow;
                    } else {
                        int ny = smoothed[ii + 1];
                        if (y > ny + halfHigh) y = ny + halfHigh;
                        if (y < ny - halfHigh) y = ny - halfHigh;
                        int n2 = smoothed[ii + 2];
                        int hi = n2 + step2;
                        int lo = n2 - step2;
                        if (y > hi) y = hi;
                        if (y < lo) y = lo;
                    }
                    smoothed[ii] = y;
                }
            }
        }

        List<Integer> out = new ArrayList<>(n);
        for (int v : smoothed) out.add(v);
        return out;
    }
}

File: common/src/main/java/net/shiroha233/roadweaver/features/path/pathlogic/core/Road.kt
package net.shiroha233.roadweaver.features.path.pathlogic.core

import net.minecraft.core.BlockPos
import net.minecraft.server.level.ServerLevel
import net.minecraft.util.RandomSource
import net.minecraft.world.level.block.Blocks
import net.minecraft.world.level.block.state.BlockState
import net.shiroha233.roadweaver.config.ConfigService
import net.shiroha233.roadweaver.config.ModConfig
import net.shiroha233.roadweaver.config.PresetService
import net.shiroha233.roadweaver.config.RoadGenerationConfig
import net.shiroha233.roadweaver.features.path.config.PathFeatureConfig
import net.shiroha233.roadweaver.features.path.pathlogic.pathfinding.RoadPathCalculator
import net.shiroha233.roadweaver.features.path.pathlogic.pathfinding.TerrainSamplingCache
import net.shiroha233.roadweaver.helpers.Records
import net.shiroha233.roadweaver.persistence.sharded.RoadShardStorage
import net.shiroha233.roadweaver.structures.precompute.RoadsideStructurePrecomputer

/**
 * 道路生成器
 *
 * 职责：根据结构连接生成一条道路，包括寻路、路面生成、路边结构预计算
 *
 * 设计原则：
 * - 接受配置快照，不依赖全局单例
 * - 所有配置在入口层读取，通过参数传递
 */
class Road(
    private val level: ServerLevel,
    private val connection: Records.StructureConnection,
    private val featureConfig: PathFeatureConfig,
    private val genConfig: RoadGenerationConfig
) {

    /**
     * 兼容旧 API：从全局配置创建
     * @deprecated 使用带 RoadGenerationConfig 参数的构造函数
     */
    @Deprecated("Use ctor with RoadGenerationConfig")
    constructor(level: ServerLevel, connection: Records.StructureConnection, config: PathFeatureConfig) :
        this(level, connection, config, RoadGenerationConfig.from(ConfigService.get()))

    /**
     * 生成道路
     *
     * @param maxSteps 最大寻路步数
     */
    fun generateRoad(maxSteps: Int) {
        val random = RandomSource.create()
        val width = genConfig.effectiveRoadWidth(getRandomWidth(random, featureConfig))

        val allowA = genConfig.allowArtificial()
        val allowN = genConfig.allowNatural()
        if (!allowA && !allowN) return

        val type = if (allowA && allowN) {
            if (random.nextBoolean()) 0 else 1
        } else {
            if (allowA) 0 else 1
        }

        val materials: List<BlockState>
        var slabMaterials: List<BlockState> = listOf()

        if (type == 0) {
            // 人工道路始终从 JSON 预设系统中选择一套材质
            val modCfg: ModConfig = ConfigService.get() // 预设服务仍需要访问全局配置
            val preset = PresetService.choosePresetForArtificial(random, modCfg)
            materials = PresetService.toBlockStatesFromIds(preset.materials())
            slabMaterials = PresetService.toBlockStatesFromIds(preset.slabMaterials())
        } else {
            materials = listOf(Blocks.DIRT_PATH.defaultBlockState(), Blocks.GRAVEL.defaultBlockState())
        }

        val rawStart = connection.from
        val rawEnd = connection.to

        // 直接用原始端点做 A* 寻路，不预设偏移方向
        val cache = TerrainSamplingCache()
        try {
            val rawSegments = RoadPathCalculator.calculateAStarRoadPath(rawStart, rawEnd, width, level, maxSteps, cache, genConfig)
            if (rawSegments == null || rawSegments.size < 5) return

            // 寻路完成后，根据实际路径方向裁剪掉进入结构保护区的路段
            // 这样即使路径从意外方向绕过来，也不会穿过结构
            val segments = StructureRoadOffsetService.trimPathNearStructure(level, rawSegments, rawStart, rawEnd)
            if (segments == null || segments.size < 5) return

            val spans = RoadPathCalculator.extractSpans(segments, level, cache, genConfig.pathfinding())
            val targetY = computeTargetY(level, segments, spans, cache, genConfig)

            val rd = Records.RoadData(width, type, materials, slabMaterials, segments, spans, targetY)
            RoadShardStorage.addRoad(level, rd)

            // 寻路完成后，预计算路边结构位置
            // 如果区块还没生成，结构会在 STRUCTURE_STARTS 阶段注入，Beardifier 会自动处理地形
            // 如果区块已经生成，则在 Feature 阶段通过 RoadsideStructurePlacer 放置（无地形适应）
            RoadsideStructurePrecomputer.precomputeStructures(level, segments, spans, width, cache, random)
        } finally {
            // 单条道路生成结束后清空噪声采样缓存，避免长时间占用内存
            cache.clear()
        }
    }

    companion object {
        private fun getRandomWidth(rnd: RandomSource, cfg: PathFeatureConfig): Int {
            return 3
        }

        private fun computeTargetY(
            level: ServerLevel,
            segments: List<Records.RoadSegmentPlacement>,
            spans: List<Records.RoadSpan>?,
            cache: TerrainSamplingCache,
            cfg: RoadGenerationConfig
        ): List<Int> {
            val n = segments.size
            val centers = ArrayList<BlockPos>(n)
            for (s in segments) centers.add(s.middlePos)

            // 将 spans 映射到索引范围，用于 BRIDGE
            val isBridge = BooleanArray(n)
            if (!spans.isNullOrEmpty()) {
                val indexMap = HashMap<Long, Int>()
                for (i in centers.indices) indexMap[centers[i].asLong()] = i

                for (sp in spans) {
                    if (sp.type != Records.SpanType.BRIDGE) continue
                    val si = indexMap[sp.start.asLong()]
                    val ei = indexMap[sp.end.asLong()]
                    if (si == null || ei == null) continue

                    val a = maxOf(0, minOf(si, ei))
                    val b = minOf(n - 1, maxOf(si, ei))
                    for (k in a..b) isBridge[k] = true
                }
            }

            val avg = maxOf(0, cfg.averagingRadius())
            val base = IntArray(n)
            for (i in 0 until n) {
                var sum = 0
                var cnt = 0
                val lo = maxOf(0, i - avg)
                val hi = minOf(n - 1, i + avg)
                for (j in lo..hi) {
                    val sp = centers[j]
                    val yTop = cache.height(level, sp.x, sp.z)
                    sum += yTop
                    cnt++
                }
                base[i] = if (cnt > 0) kotlin.math.round(sum.toDouble() / cnt.toDouble()).toInt() else centers[i].y
            }

            // 如果关闭限坡平滑，则直接使用基础平均高度，不再进行每两段步进限制
            if (!cfg.slopeLimitEnabled()) {
                return base.toList()
            }

            val smoothed = base.clone()
            // 对每个连续非桥梁段进行平滑，以避免奇偶振荡
            var i = 0
            while (i < n) {
                while (i < n && isBridge[i]) i++
                val s = i
                while (i < n && !isBridge[i]) i++
                val e = i - 1

                if (s <= e) {
                    val step2 = maxOf(0, minOf(8, cfg.maxSlopeStepPerTwoSegments()))
                    val halfLow = maxOf(0, step2 / 2)
                    val halfHigh = maxOf(0, (step2 + 1) / 2)

                    for (ii in (s + 1)..e) {
                        var y = smoothed[ii]
                        if (ii == s + 1) {
                            val py = smoothed[ii - 1]
                            if (y > py + halfLow) y = py + halfLow
                            if (y < py - halfLow) y = py - halfLow
                        } else {
                            val py = smoothed[ii - 1]
                            if (y > py + halfHigh) y = py + halfHigh
                            if (y < py - halfHigh) y = py - halfHigh

                            val p2 = smoothed[ii - 2]
                            val hi = p2 + step2
                            val lo = p2 - step2
                            if (y > hi) y = hi
                            if (y < lo) y = lo
                        }
                        smoothed[ii] = y
                    }

                    for (ii in (e - 1) downTo s) {
                        var y = smoothed[ii]
                        if (ii == e - 1) {
                            val ny = smoothed[ii + 1]
                            if (y > ny + halfLow) y = ny + halfLow
                            if (y < ny - halfLow) y = ny - halfLow
                        } else {
                            val ny = smoothed[ii + 1]
                            if (y > ny + halfHigh) y = ny + halfHigh
                            if (y < ny - halfHigh) y = ny - halfHigh

                            val n2 = smoothed[ii + 2]
                            val hi = n2 + step2
                            val lo = n2 - step2
                            if (y > hi) y = hi
                            if (y < lo) y = lo
                        }
                        smoothed[ii] = y
                    }
                }
            }

            return smoothed.toList()
        }
    }
}
